### PR TITLE
Don't reuse other syntax nodes for the with expression

### PR DESCRIPTION
**Pull request by agocke**
The semantic model sees AnonymousObjectCreation nodes as all belonging
to anonymous object creation, which causes asserts and bad casts when
that's not true. For now it's best to just use simple nodes in the
WithExpression.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -10610,6 +10592,84 @@ private bool IsTargetTypedObjectCreation()
...
+                            }
+                            assignments.Add(parseAssignment());
+                        }
+                        else if (skipBadTokens(ref openBrace, SyntaxKind.CommaToken) == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+                    }
+                }
+                else if (skipBadTokens(ref openBrace, SyntaxKind.IdentifierToken) == PostSkipAction.Continue)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10629]: >skipBadTokens [](start = 25, length = 13)

I'm afraid this will result in pretty bad error recovery.  I suggest instead simply parsing an expression.  If it ends up being an assignment expression, use that.  If not, add a missing "identifier equals" with a syntax error and use that.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -10610,6 +10592,84 @@ private bool IsTargetTypedObjectCreation()
...
         }
 
+#nullable enable
+
+        private ExpressionSyntax ParseWithExpression(ExpressionSyntax receiverExpression, SyntaxToken withKeyword)
+        {
+            var openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
+
+            var assignments = _pool.AllocateSeparated<AssignmentExpressionSyntax>();
+            if (CurrentToken.Kind != SyntaxKind.CloseBraceToken)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10602]: >if [](start = 12, length = 2)

I think this should probably be a while. the `goto` is a bit hard to follow.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -10610,6 +10592,84 @@ private bool IsTargetTypedObjectCreation()
...
+                    SyntaxKind.SimpleAssignmentExpression,
+                    identifier,
+                    equal,
+                    expression);
+            }
+
+            PostSkipAction skipBadTokens(ref SyntaxToken startToken, SyntaxKind expected)
+            {
+                return this.SkipBadSeparatedListTokensWithExpectedKind(ref startToken, assignments,
+                    p => p.CurrentToken.Kind != SyntaxKind.IdentifierToken && !p.IsNamedAssignment(),
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10665]: >IsNamedAssignment [](start = 81, length = 17)

If the token kind is not identifier, then IsNamedAssignment always returns false. In other words, I think the second half of this `&&` operation does not contribute the the result.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 4]
```diff
@@ -460,11 +460,21 @@
     </Field>
     <Field Name="OpenBraceToken" Type="SyntaxToken">
       <Kind Name="OpenBraceToken"/>
+      <PropertyComment>
+        <summary>SyntaxToken representing the open brace.</summary>
+      </PropertyComment>
+    </Field>
+    <!-- PROTOTYPE: We could try to reuse the InitializerExpressionSyntax here. -->
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:467]: wouldn't it be cheaper to just do that now and gate in binding?

**Comment by jcouv** [comment]
I understand that there were problems with re-using syntax from anonymous types. But I wonder if we should be re-using syntax from object creation initializers. It's true that they support some initialization which is not supported in `with` expressions (for example `{ [0] = 42 }`, but they actually seem a good fit and desirable (maybe we would want to support them at some point).

I'm proposing:
``` antlr
with_expression
  : expression 'with' initializer_expression
  ;
```

**Comment by CyrusNajmabadi** [comment]
i agree with @jcouv  :)

**Comment by agocke** [comment]
Works for me. I'll turn this back into draft until I can update.

**Comment by alrz** [comment]
>expression 'with' initializer_expression

Also it'll work with collections and nested forms (https://github.com/dotnet/csharplang/issues/3262), assuming the syntax is acceptable. :)

**Comment by agocke** [comment]
This should be ready for re-review, I've updated parsing to look for any expressions and reuse the InitializerExpression nodes. Now we validate that each expression is an assignment in binding.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 5]
```diff
@@ -1519,6 +1477,45 @@
...
+    <Field Name="WithKeyword" Type="SyntaxToken">
+      <Kind Name="WithKeyword" />
+    </Field>
+    <Field Name="Initializer" Type="InitializerExpressionSyntax">
+        <PropertyComment>
+            <summary>InitializerExpressionSyntax representing the initializer expression for the with expression.</summary>
+        </PropertyComment>
+    </Field>
+  </Node>
+  <Node Name="MemberAccessExpressionSyntax" Base="ExpressionSyntax">
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:1492]: >MemberAccessExpressionSyntax [](start = 14, length = 28)

This has moved in the .xml file for no apparent reason.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt [hunk 6]
```diff
@@ -1,3 +1,9 @@
+*REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.ArgumentList.get -> Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentListSyntax
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:1]: >ObjectCreationExpressionSyntax [](start = 46, length = 30)

Why are there changes here? I don't see any change to the syntax node definitions.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 7]
```diff
@@ -782,7 +785,10 @@ public static void Main()
             comp.VerifyDiagnostics(
                 // (8,22): error CS1525: Invalid expression term '='
                 //         c = c with { = 5 };
-                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "=").WithArguments("=").WithLocation(8, 22)
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "=").WithArguments("=").WithLocation(8, 22),
+                // (8,22): error CS1061: 'C' does not contain a definition for '' and no accessible extension method '' accepting a first argument of type 'C' could be found (are you missing a using directive or an assembly reference?)
+                //         c = c with { = 5 };
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "").WithArguments("C", "").WithLocation(8, 22)
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:791]: >ERR_NoSuchMemberOrExtension [](start = 37, length = 27)

This feels like a cascaded diagnostic. Please fix it or take a note of that for possible future improvement.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 6 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:1]: Hmm, those are strange. I'll investigate.

Comment (agocke) [reply src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:1]: OK, figured it out, these are just re-ordered because I ran the sorting script to resolve merge conflicts.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 7 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:791]: Yup, I have a follow-up PR that should address it.

File: src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4 [hunk 8]
```diff
@@ -996,7 +996,7 @@ type_of_expression
   ;
 
 with_expression
-  : expression 'with' '{' (anonymous_object_member_declarator (',' anonymous_object_member_declarator)* ','?)? '}'
+  : expression 'with' initializer_expression
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4:999]: noice.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 9]
```diff
@@ -10735,6 +10717,77 @@ private bool IsTargetTypedObjectCreation()
...
+                openBrace,
+                _pool.ToListAndFree(list),
+                closeBrace);
+
+            withKeyword = CheckFeatureAvailability(withKeyword, MessageID.IDS_FeatureRecords);
+
+            var result = _syntaxFactory.WithExpression(
+                receiverExpression,
+                withKeyword,
+                initializer);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:10784]:
```suggestion
            var result = _syntaxFactory.WithExpression(receiverExpression, withKeyword, initializer);
```
?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 10]
```diff
@@ -1427,5 +1432,73 @@ public static void Main()
...
+        [Fact]
+        public void WithSemanticModel1()
+        {
+            var src = @"
+data class C(int X, string Y)
+{
+    public static void Main()
+    {
+        var c = new C(0, ""a"");
+        c = c with { X = 2 };
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:1445]: > with { X = 2 } [](start = 13, length = 15)

Missed the window for this PR, but consider adding tests for other expressions that are used in different initializers but disallowed in `with` initializers. For example `[0] = 2` or values for `IEnumerable/Add` collection.
